File: trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        Deque<int[]> stack = new LinkedList<>();
        
        int result = 0;
        
        for(int i = 0; i < height.length; i++) {
            int h = height[i];
            
            while(!stack.isEmpty() && stack.peek()[1] < h) {
                int[] prev = stack.pop();
                
                if (stack.isEmpty()) continue;
                
                int rh = Math.min(h, stack.peek()[1]) - prev[1];
                int w = i - stack.peek()[0] - 1;
                
                result += Math.max(0, rh * w);
            }
            
            stack.push(new int[] {i, h});
        }
        
        return result;
    }
}

File: trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []
        
        res = 0
        for i, h in enumerate(height):
            while len(stack) > 0 and stack[-1][0] <= h:
                tmp = stack.pop()
                p = (0, -1) if len(stack) == 0 else stack[-1]
                res += max(0, (min(h, p[0]) - tmp[0]) * (i - p[1] - 1))
            stack.append((h, i))
            
        return res
        
            